File: experiments/BENEFICIAL_GENESIS_REPRODUCER_001/src/encoding.mjs
/**
 * Canonical encodings for Beneficial Genesis (clean-room Node implementation).
 *
 * domain_hash recovered from fixtures:
 *   SHA-256( domain || 0x00 || u32be(len(p)) || p  for each part )
 *
 * Donation commitment uses the special string-separated preimage from
 * TECHNICAL_DESIGN.md §4.2 (not length-prefixed domain_hash).
 */

import crypto from "node:crypto";

export function sha256(buf) {
  return crypto.createHash("sha256").update(buf).digest();
}

export function sha256Hex(buf) {
  return sha256(buf).toString("hex");
}

export function hmacSha256(key, data) {
  return crypto.createHmac("sha256", key).update(data).digest();
}

export function u32Be(n) {
  if (typeof n !== "number" || !Number.isInteger(n) || n < 0 || n > 0xffffffff) {
    throw new TypeError("u32 out of range");
  }
  const b = Buffer.alloc(4);
  b.writeUInt32BE(n >>> 0);
  return b;
}

export function u64Be(n) {
  if (typeof n !== "number" || !Number.isInteger(n) || n < 0 || n > Number.MAX_SAFE_INTEGER) {
    // still allow up to 2^53-1 safely; larger values rejected as TYPE_ERROR upstream
    if (typeof n !== "number" || !Number.isInteger(n) || n < 0) {
      throw new TypeError("u64 out of range");
    }
  }
  const b = Buffer.alloc(8);
  b.writeBigUInt64BE(BigInt(n));
  return b;
}

/** Recursively sort object keys; match Python json.dumps(sort_keys=True, separators=(',',':'), ensure_ascii=True). */
export function sortKeys(obj) {
  if (Array.isArray(obj)) return obj.map(sortKeys);
  if (obj && typeof obj === "object") {
    const out = {};
    for (const k of Object.keys(obj).sort()) out[k] = sortKeys(obj[k]);
    return out;
  }
  return obj;
}

export function canonicalJson(obj) {
  // JSON.stringify escapes non-ASCII as \uXXXX when needed similarly for our fixture alphabet.
  return JSON.stringify(sortKeys(obj));
}

export function canonicalJsonBytes(obj) {
  return Buffer.from(canonicalJson(obj), "utf8");
}

/**
 * domain_hash(domain, *parts) -> bytes
 * SHA-256 over domain || 0x00 || len_be32(part) || part for each part.
 */
export function domainHash(domain, ...parts) {
  const chunks = [domain, Buffer.from([0x00])];
  for (const part of parts) {
    if (!Buffer.isBuffer(part)) {
      throw new TypeError("domain_hash parts must be Buffer");
    }
    chunks.push(u32Be(part.length), part);
  }
  return sha256(Buffer.concat(chunks));
}

export function domainHashHex(domain, ...parts) {
  return domainHash(domain, ...parts).toString("hex");
}

/**
 * Strict lowercase hex parse. Throws TypeError/ValueError-like Error with message.
 * @param {string} name
 * @param {unknown} value
 * @param {{ expectedBytes?: number }} [opts]
 * @returns {Buffer}
 */
export function requireHex(name, value, opts = {}) {
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be hex string`);
  }
  if (value.length === 0) {
    throw new TypeError(`${name} empty hex`);
  }
  if (value.length % 2 !== 0) {
    throw new TypeError(`${name} odd-length hex`);
  }
  if (!/^[0-9a-f]+$/.test(value)) {
    // reject uppercase / non-hex
    throw new TypeError(`${name} malformed hex`);
  }
  const buf = Buffer.from(value, "hex");
  if (opts.expectedBytes !== undefined && buf.length !== opts.expectedBytes) {
    throw new TypeError(`${name} expected ${opts.expectedBytes} bytes`);
  }
  return buf;
}

/** True integer check — rejects boolean (typeof true === 'boolean' in JS; also reject Number objects). */
export function isStrictInt(value) {
  return typeof value === "number" && Number.isInteger(value) && !Object.is(value, -0) ||
    (typeof value === "number" && Number.isInteger(value));
}

/** Reject bool and non-integers. */
export function asNonNegInt(name, value, max) {
  // In JS, typeof true === 'boolean', so bool is already excluded from number.
  // Still reject NaN and non-integers.
  if (typeof value === "boolean") {
    throw new TypeError(`${name} must be int not bool`);
  }
  if (typeof value !== "number" || !Number.isInteger(value)) {
    throw new TypeError(`${name} must be int`);
  }
  if (value < 0 || value > max) {
    throw new TypeError(`${name} out of range`);
  }
  return value;
}

File: experiments/BENEFICIAL_GENESIS_DESIGN_001/protocol/encoding.py
"""Canonical encoding helpers for Beneficial Genesis objects.

Rules:
- Canonical JSON: UTF-8, sorted object keys, no insignificant whitespace,
  separators=(',', ':'), ensure_ascii=False for readability of test labels only
  in human docs; protocol digests use ensure_ascii=True so codepoints are stable.
- Integers are non-negative Python ints within declared widths.
- Binary fields are lowercase hex without 0x prefix.
- Multi-byte integers in commitment preimages are big-endian.
"""

from __future__ import annotations

import hashlib
import json
import struct
from typing import Any, Mapping


def sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def sha256_hex(data: bytes) -> str:
    return sha256(data).hex()


def double_sha256(data: bytes) -> bytes:
    return sha256(sha256(data))


def double_sha256_hex(data: bytes) -> str:
    return double_sha256(data).hex()


def u32_be(value: int) -> bytes:
    if not isinstance(value, int) or value < 0 or value > 0xFFFFFFFF:
        raise ValueError("u32 out of range")
    return struct.pack(">I", value)


def u64_be(value: int) -> bytes:
    if not isinstance(value, int) or value < 0 or value > 0xFFFFFFFFFFFFFFFF:
        raise ValueError("u64 out of range")
    return struct.pack(">Q", value)


def require_hex(name: str, value: str, *, expected_bytes: int | None = None) -> bytes:
    if not isinstance(value, str):
        raise TypeError(f"{name} must be hex string")
    if len(value) % 2 != 0 or any(c not in "0123456789abcdef" for c in value):
        raise ValueError(f"{name} must be lowercase hex")
    raw = bytes.fromhex(value)
    if expected_bytes is not None and len(raw) != expected_bytes:
        raise ValueError(f"{name} must be exactly {expected_bytes} bytes")
    return raw


def canonical_json_bytes(obj: Any) -> bytes:
    """Return deterministic JSON bytes for hashing and fixture equality."""

    return json.dumps(
        obj,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=True,
        allow_nan=False,
    ).encode("utf-8")


def canonical_json_loads(text: str | bytes) -> Any:
    if isinstance(text, bytes):
        text = text.decode("utf-8")
    return json.loads(text)


def deep_sorted(obj: Any) -> Any:
    """Recursively sort mapping keys for defensive comparisons."""

    if isinstance(obj, Mapping):
        return {k: deep_sorted(obj[k]) for k in sorted(obj)}
    if isinstance(obj, list):
        return [deep_sorted(x) for x in obj]
    return obj


def domain_hash(domain: bytes, *parts: bytes) -> bytes:
    """SHA-256 over domain || 0x00 || len(part)||part for each part."""

    acc = bytearray(domain)
    acc.append(0x00)
    for part in parts:
        if not isinstance(part, (bytes, bytearray)):
            raise TypeError("domain_hash parts must be bytes")
        acc.extend(u32_be(len(part)))
        acc.extend(part)
    return sha256(bytes(acc))


def domain_hash_hex(domain: bytes, *parts: bytes) -> str:
    return domain_hash(domain, *parts).hex()
